docs/docs/async-dialogs.md

```diff
@@ -4,3 +4,628 @@ title: Async Dialogs
+
+  return new Promise(function(resolve, reject) {
+    resolver = resolve;
+    rejector = reject;
+  })
+  .finally(function() {
+    okButton.removeEventListener('click', handleOk);
+    cancelButton.removeEventListener('click', handleCancel);
+    dialog.className = 'hidden';
+  });
+}
+```
+
+[Run example on JSBin](http://jsbin.com/fucofu/edit?js,output)
+
+Now when the user presses the **Set Name** button the clock continues to update
+while the dialog is visible.
+
+Because the `removeEventListener` requires a reference to the original function
+that was used with the `addEventListener` it makes it difficult to clean up
+after itself without storing the references in a scope higher then the promise
+resolution function. This pattern of storing the `resolve` and `reject`
+callbacks outside the scope of the promise function is called [the **deferred**
+anti-pattern](api/anti-patterns.html#the-deferred-anti-pattern). It is an
+*anti-pattern* because it is often used when it is not needed. However, in this
+case it is the only way to accomplish the clean-up requirements imposed by the
+DOM API (namely the need for the original function reference to
+`removeEventListener`).
+
+If this was using jQuery there would be no need for the deferred pattern because
+jQuery can remove events handlers without the need for a function reference
+using [event namespacing](https://api.jquery.com/on/#event-names).
+
+```javascript
+return new Promise(function(resolve, reject) {
+  $('#okButton').on('click.promptDialog', function() {
+    resolve(input.value);
+  });
+  $('#cancelButton').on('click.promptDialog', reject);
+})
+.finally(function() {
+  $('#okButton').off('click.promptDialog');
+  $('#cancelButton').off('click.promptDialog');
+});
+```
+
+Now looking at the example above there are a few issues with it. Mainly the need
+for the deferred anti-pattern and it feels like it is doing too much. A *squint*
+test reveals four inner functions, six variable references, and its behavior is
+to show the dialog, attach two DOM events, construct a promise, hide the dialog,
+and finally detach DOM events. That is a lot for one little function. A
+refactoring can help.
+
+Abstraction is the key here. We will make an *object* (or class) that is
+responsible for managing the dialog box. Its interface will manage only two
+function references (callbacks): when the user clicks ok and when user clicks
+cancel. And it will offer the value when asked.
+
+Using an abstraction like this the `promisePrompt` no longer needs to know
+anything about the DOM and concentrates on just providing a promise. This will
+also make things easier to create a promised version of a progress bar or
+confirmation dialog or any other type of UI that we want to have a value for.
+All we will need to do is write a class for that dialog type with the same
+interface and just pass that class into our promise making method.
+
+The dialog interface might look like this:
```

Comment:
I think it's starting to lose a little focus here. You're introducing a class without an obvious immediate benefit.